kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/punctuators/AbstractThrottledPunctuator.java

```diff
@@ -0,0 +1,97 @@
+package org.hypertrace.core.kafkastreams.framework.punctuators;
+
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.hypertrace.core.kafkastreams.framework.punctuators.action.TaskResult;
+
+@Slf4j
+public abstract class AbstractThrottledPunctuator<T> implements Punctuator {
+  private final Clock clock;
+  private final KeyValueStore<Long, ArrayList<T>> objectStore;
+  private final ThrottledPunctuatorConfig config;
+
+  public AbstractThrottledPunctuator(
+      Clock clock,
+      ThrottledPunctuatorConfig config,
+      KeyValueStore<Long, ArrayList<T>> objectStore) {
+    this.clock = clock;
+    this.config = config;
+    this.objectStore = objectStore;
+  }
+
+  public void scheduleTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    objectsAtWindow.add(object);
+    objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+  }
+
+  public boolean rescheduleTask(long atTimestampInMs, long toTimestampInMs, T object) {
+    if (cancelTask(atTimestampInMs, object)) {
+      scheduleTask(toTimestampInMs, object);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  public boolean cancelTask(long atTimestampInMs, T object) {
+    long windowAlignedTimestamp = getWindowAlignedTimestamp(atTimestampInMs);
+    ArrayList<T> objectsAtWindow =
+        Optional.ofNullable(objectStore.get(windowAlignedTimestamp)).orElse(new ArrayList<>());
+    boolean removed = objectsAtWindow.remove(object);
+    if (objectsAtWindow.isEmpty()) {
+      objectStore.delete(windowAlignedTimestamp);
+    } else {
+      objectStore.put(windowAlignedTimestamp, objectsAtWindow);
+    }
+    return removed;
+  }
+
+  @Override
+  public final void punctuate(long punctuateTimestamp) {
+    long startTimestamp = clock.millis();
+    log.debug(
+        "Processing tasks with throttling yield of {} until timestamp {}",
+        config.getYieldMs(),
+        punctuateTimestamp);
+    try (KeyValueIterator<Long, ArrayList<T>> it = objectStore.range(0L, punctuateTimestamp)) {
```

Comment:
You are iterating through keys, and will call executeTask. In the executeTask call, the implementation can reschedule task. This is like updating the iterator during the iteration. Truly a case of ConcurrentModificationException ;-)

Reply:
The iterator provided is safe against ConcurrentModificationException,  I can modify the task to collection first and delete later. But that will introduce a point of failure as you mentioned similar to other comment.

Although I'm not clear on how the iterator is implemented safely and is it desirable to use it in concurrent modification way. Couldn't figure it out, please let me know if you have an explanation

Reply:
There are two options -
1. Option one - Remove cancel impl, and then delete all handled keys(windowedTs) after closing the iterator.
2. Option two - Make a copy of key and values in a separate map, and close this iterator. Then iterate over the copy, and not this iterator.

Reply:
But given it is safe against ConcurrentModificationException by design, do we need to resort to one of these two solutions?

Reply:
Looked at the range(...) javadoc. This should be fine.